x-pack/plugins/remote_clusters/server/routes/api/remote_clusters/register_delete_route.js

```diff
@@ -20,43 +20,58 @@ export function registerDeleteRoute(server) {
   server.route({
     path: '/api/remote_clusters/{name}',
     method: 'DELETE',
+    config: {
+      pre: [ licensePreRouting ]
+    },
     handler: async (request) => {
       const callWithRequest = callWithRequestFactory(server, request);
       const { name } = request.params;
+      const names = name.split(',');
 
-      // Check if cluster does exist
-      try {
-        const existingCluster = await doesClusterExist(callWithRequest, name);
-        if(!existingCluster) {
-          return wrapCustomError(new Error('There is no remote cluster with that name.'), 404);
-        }
-      } catch (err) {
-        return wrapCustomError(err, 400);
-      }
-
-      try {
-        const deleteClusterPayload = serializeCluster({ name });
-        const response = await callWithRequest('cluster.putSettings', { body: deleteClusterPayload });
-        const acknowledged = get(response, 'acknowledged');
-        const cluster = get(response, `persistent.cluster.remote.${name}`);
-
-        if (acknowledged && !cluster) {
-          return { success: true };
-        }
-
-        // If for some reason the ES response still returns the cluster information,
-        // return an error. This shouldn't happen.
-        return wrapCustomError(new Error('Unable to delete cluster, information still returned from ES.'), 400);
-      } catch (err) {
-        if (isEsError(err)) {
-          return wrapEsError(err);
-        }
-
-        return wrapUnknownError(err);
-      }
-    },
-    config: {
-      pre: [ licensePreRouting ]
+      const itemsDeleted = [];
+      const errors = [];
+
+      await Promise.all(names.map((clusterName) => (
```

Comment:
I think it would be clearer to extract the function and give it a name (like I did in rollup https://github.com/elastic/kibana/blob/master/x-pack/plugins/rollup/server/routes/api/jobs.js#L119).

I also had a hard time to read it, I would suggest to split it in parts. What do you think about this:

```js
handler: async (request) => {
  const callWithRequest = callWithRequestFactory(server, request);
  const { name } = request.params;
  const names = name.split(',');

  const itemsDeleted = [];
  const errors = [];

  // Validator that returns an error if the remote cluster does not exist
  const validateClusterDoesExist = async (name) => {
    try {
      const existingCluster = await doesClusterExist(callWithRequest, name);
      return existingCluster
        ? null
        : wrapCustomError(new Error(`The cluster ${name} does not exist.`), 404);
    } catch (err) {
      return wrapCustomError(err, 400);
    }
  };

  // Send the request to delete the cluster and returns an error if it could not be deleted
  const sendRequestToDeleteCluster = async (name) => {
    const body = serializeCluster({ name });

    try {
      const response = await callWithRequest('cluster.putSettings', { body });
      const acknowledged = get(response, 'acknowledged');
      const cluster = get(response, `persistent.cluster.remote.${name}`);

      if (acknowledged && !cluster) {
        return null;
      }

      // If for some reason the ES response still returns the cluster information,
      // return an error. This shouldn't happen.
      return wrapCustomError(new Error('Unable to delete cluster, information still returned from ES.'), 400);
    } catch(err) {
      return isEsError(err)
        ? wrapEsError(err)
        : wrapUnknownError(err);
    }
  };

  const deleteCluster = async (clusterName) => {
    try {
      // Validate that the cluster exists
      let error = await validateClusterDoesExist(clusterName);

      if (!error) {
        // Delete the cluster
        error = await sendRequestToDeleteCluster(clusterName);
      }

      if (error) {
        throw error;
      }

      // If we are here, it means that everything went well...
      itemsDeleted.push(clusterName);
    } catch(err) {
      errors.push({ name: clusterName, error: err });
    }
  };

  // Delete all our cluster in parallel
  await Promise.all(names.map(deleteCluster));

  return {
    itemsDeleted,
    errors,
  };
}
```

Reply:
I like this much better! Thanks for the great suggestion.